aws_lambda_powertools/utilities/batch/base.py

```diff
@@ -146,3 +185,120 @@ def batch_processor(
         processor.process()
 
     return handler(event, context)
+
+
+class BatchProcessor(BasePartialProcessor):
```

Comment:
would it make sense to have separate processors for each event type (SQS, DynamoDB or Kinesis) instead of growing the complexity of this class?  Then you could encapsulate the failure collection in the specific processor.

Reply:
That was the initial version we wanted to implemented - a KinesisDataStreamProcessor, DynamoDB... then @cakepietoast argued that this was gonna confuse customers with other available processors (Sqs, PartialProcessor, BaseProcessor), as we can only deprecate them in v2.

I'm 50/50 here if I'm honest. I prefer a separate one but I also can see customers easily confused of which one to pick despite docs change I'm gonna make.

Implementation wise, this will remain stable. The only two changes I can anticipate is 1/ supporting the new Permanent Exception parameter, and 2/ raising a descriptive exception in case we reach an AttributeError when collecting message id/sequence number from a malformed event/model.

Reply:
What if we changed the nomenclature to be “producer” and “consumer” (these processors would be consumers). I had that other idea earlier to make it easier to use the SQS and DynamoDB batch write methods taking into account their batch sizes, those could be “producers” 🤷‍♂️

Reply:
Where would the producer sit and what would be its responsibilities?

For that suggestion on partitioning, we **should** add it to the Event Source Data Class as it's a no brainier.

I **think** the word Consumer wouldn't be explicit enough on the capabilities Batch provide - maybe something else then?

Features

- Transform incoming records into Event Source Data Class or Pydantic Models
- Call an user defined function for each record in the batch
- Keep track of exceptions raised for any record
- For partial successes, extract message IDs (SQS) or sequence numbers (DynamoDB & Kinesis) and build a response to the new BatchItemIdentifiers contract
- Future: If a raised exception  matches the permanent_exceptions set, send these records to the configured DLQ in batches of 10